core/platform/auth/firebase_auth_services_test.py

```diff
@@ -0,0 +1,270 @@
+            with firebase_auth_services.acquire_auth_context():
+                context_enterred = True
+        self.assertTrue(context_enterred)
+
+    def test_verify_accepted_token(self):
+        with self.swap_verify_to_always_accept_tokens(sub='123'):
+            self.assertEqual(
+                firebase_auth_services.get_verified_sub(
+                    self.make_response(auth_header='Bearer MOCK_JWT_VALUE')),
+                '123')
+
+    def test_verify_rejected_token(self):
+        err = Exception('untrusted!')
+        with self.swap_verify_to_always_reject_tokens(exception_obj=err):
+            with self.assertRaisesRegexp(Exception, 'untrusted!'):
+                firebase_auth_services.get_verified_sub(
+                    self.make_response(auth_header='Bearer MOCK_JWT_VALUE'))
+
+    def test_verify_token_without_auth_header(self):
+        with self.swap_verify_to_always_accept_tokens():
+            with self.assertRaisesRegexp(Exception, 'header is missing'):
+                firebase_auth_services.get_verified_sub(
+                    self.make_response(auth_header=None))
+
+    def test_verify_token_with_empty_auth_header(self):
+        with self.swap_verify_to_always_accept_tokens():
+            with self.assertRaisesRegexp(Exception, 'prefix is missing'):
+                firebase_auth_services.get_verified_sub(
+                    self.make_response(auth_header=''))
+
+    def test_verify_token_with_auth_header_missing_bearer_prefix(self):
+        with self.swap_verify_to_always_accept_tokens():
+            with self.assertRaisesRegexp(Exception, 'prefix is missing'):
+                firebase_auth_services.get_verified_sub(
+                    self.make_response(auth_header='JWT_TOKEN'))
+
+    def test_verify_token_with_auth_header_missing_space_after_bearer(self):
+        with self.swap_verify_to_always_accept_tokens():
+            with self.assertRaisesRegexp(Exception, 'prefix is missing'):
+                firebase_auth_services.get_verified_sub(
+                    self.make_response(auth_header='BearerJWT_TOKEN'))
+
+
+class UserIdByFirebaseSubModelTests(test_utils.GenericTestBase):
+    """Tests for _UserIdByFirebaseSubModel."""
+
+    NONEXISTENT_AUTH_METHOD_NAME = 'auth_method_x'
+    NONEXISTENT_USER_ID = 'id_x'
+    NONREGISTERED_GAE_ID = 'gae_id_x'
+    USER_ID = 'user_id'
+    USER_GAE_ID = 'gae_id'
+    PROFILE_ID = 'profile_id'
+    PROFILE_2_ID = 'profile_2_id'
+
+    def setUp(self):
+        """Set up user models in datastore for use in testing."""
+        super(UserIdByFirebaseSubModelTests, self).setUp()
+
+        firebase_auth_services._UserIdByFirebaseSubModel( # pylint: disable=protected-access
+            id=self.USER_GAE_ID, user_id=self.USER_ID).put()
+
+    def test_get_deletion_policy_is_delete_at_end(self):
+        self.assertEqual(
+            firebase_auth_services._UserIdByFirebaseSubModel # pylint: disable=protected-access
```

Comment:
It's the first time I've seen us declare a private model class. Why is this necessary?

Reply:
Yeah I changed my mind immediately after writing it (while I was trying to fall asleep :P).

Changed.